compile to native backend failed, use msvc

Original: https://github.com/moonbitlang/moonbit-docs/issues/445

When use pointer arithmetic, use char* rather than  void*

[error C2036: 'void *' : unknown size](https://stackoverflow.com/a/37460703/22152846)

PS C:\illu\repo\dev-mbt\memory.mbt> moon test --target=native
[===================================-----] 7/8 done, 1/1 running
link-core: tonyfettes/memory_blackbox_test
  Microsoft (R) C/C++ Optimizing Compiler Version 19.42.34436 for x86
Microsoft (R) C/C++ Optimizing Compiler Version 19.42.34436 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

cl : Command line warning D9035 : option 'o' has been deprecated and will be removed in a future release
cl : Command line warning D9002 : ignoring unknown option '-fno-strict-aliasing'
memory.blackbox_test.c
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14477): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14484): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14493): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14500): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14509): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14516): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14525): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14532): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14541): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14548): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14557): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14564): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14573): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14580): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14620): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14629): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14638): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14647): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14656): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14665): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14674): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14731): error C2036: 'void *': unknown size
C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.c(14739): error C2036: 'void *': unknown size
failed to execute ' C:\illu\repo\dev-mbt\memory.mbt\target\native\debug\test\memory.blackbox_test.exe'

Caused by:
    The system cannot find the file specified. (os error 2)

Total tests: 14, passed: 0, failed: 14.

--- a/src/malloc.native.mbt
+++ b/src/malloc.native.mbt
@@ -12,7 +12,11 @@ extern "c" fn _moonbit_decref(ptr : Memory[Unit]) = "moonbit_decref"
 
 ///|
 pub fn free[T](ptr : Memory[T]) -> Unit {
-  _moonbit_decref(ptr.cast().sub(8).cast())
+  ptr.cast()
+  |> _array_of_memory()
+  |> _array_sub(1)
+  |> _array_to_memory()
+  |> _moonbit_decref()
 }
 
 ///|

--- a/src/memory.native.mbt
+++ b/src/memory.native.mbt
@@ -25,41 +25,46 @@ fn Memory::cast[T, U](self : Memory[T]) -> Memory[U] = "%identity"
 ///|
 /// This will be compiled as `+` and will be inlined. This way we can bypass the
 /// typecheck on the C side.
-fn Memory::add(self : Memory[Unit], offset : Int) -> Memory[Unit] = "%u64.add"
+fn _array_add[T](array : FixedArray[T], offset : Int) -> FixedArray[T] = "%u64.add"
 
 ///|
-fn Memory::sub(self : Memory[Unit], offset : Int) -> Memory[Unit] = "%u64.sub"
+fn _array_sub[T](array : FixedArray[T], offset : Int) -> FixedArray[T] = "%u64.sub"
+
+///|
+fn _string_add(string : String, offset : Int) -> String = "%u64.add"
+
+///|
+fn _string_sub(string : String, offset : Int) -> String = "%u64.sub"
 
 ///|
 pub fn Memory::to_array[T : Sized + Default + Get + Set](
   self : Memory[T]
 ) -> FixedArray[T] {
-  _moonbit_incref(self.cast().sub(8).cast())
-  _array_of_memory(self.cast().sub(8).cast())
+  let array = _array_sub(_array_of_memory(self), 1)
+  _moonbit_incref(_array_to_memory(array).cast())
+  array
 }
 
 ///|
 pub fn Memory::of_array[T : Sized + Default + Set](
   array : FixedArray[T]
 ) -> Memory[T] {
-  let memory = _array_to_memory(array)
-  memory.cast().add(8).cast()
+  _array_to_memory(_array_add(array, 1))
 }
 
 ///|
 pub fn Memory::to_string(self : Memory[Byte]) -> String {
-  _moonbit_incref(self.cast().sub(8).cast())
-  _string_of_memory(self.cast().sub(8).cast())
+  let string = _string_sub(_string_of_memory(self), 1)
+  _moonbit_incref(_string_to_memory(string).cast())
+  string
 }
 
 ///|
 pub fn Memory::of_string(string : String) -> Memory[Byte] {
-  let memory : Memory[Byte] = _string_to_memory(string)
-  memory.cast().add(8).cast()
+  _string_to_memory(_string_add(string, 1))
 }
 
 ///|
 pub fn Memory::length[T : Sized + Default + Get + Set](self : Memory[T]) -> Int {
-  let array = self.to_array()
-  array.length()
+  self.to_array().length()
 }

--- a/src/size.mbt
+++ b/src/size.mbt
@@ -1,7 +1,6 @@
 ///|
 type Size Int
 
-
 ///|
 pub fn Size::op_get(size : Size, index : Int) -> Size = "%i32_mul"
 

